pkg/eventrecorder/eventrecorder.go

```diff
@@ -71,167 +143,71 @@ type eventReport struct {
-		evt.EventDetails = &EventDetailsIndexer{
-			CandidateCount: uint64(len(ret.Candidates())),
-			Protocols:      toStrings(ret.Protocols()),
+func batchTransform(instanceID string, retrievalEvents []types.RetrievalEvent) batchedRawEvents {
+	rawEvents := make([]rawEvent, 0, len(retrievalEvents))
+
+	for _, event := range retrievalEvents {
+		rawEvent := rawEvent{
+			RetrievalId:       event.RetrievalId(),
+			InstanceId:        instanceID,
+			Cid:               event.PayloadCid().String(),
+			StorageProviderId: types.Identifier(event),
+			Phase:             event.Phase(),
+			PhaseStartTime:    event.PhaseStartTime(),
+			EventName:         event.Code(),
+			EventTime:         event.Time(),
 		}
-	case events.RetrievalEventConnected:
-	case events.EventWithQueryResponse: // events.RetrievalEventQueryAsked, events.RetrievalEventQueryAskedFiltered:
-		qr := ret.QueryResponse()
-		evt.EventDetails = &qr
-	case events.RetrievalEventProposed:
-	case events.RetrievalEventAccepted:
-	case events.RetrievalEventFirstByte:
-	case events.RetrievalEventFailed:
-		evt.EventDetails = &EventDetailsError{ret.ErrorMessage()}
-	case events.RetrievalEventSuccess:
-		evt.EventDetails = &EventDetailsSuccess{ret.ReceivedSize(), ret.ReceivedCids(), uint64(ret.Duration().Milliseconds())}
-	}
-
-	er.recordEvent(reflect.TypeOf(event).Name(), evt)
-}
-
-func (er *EventRecorder) recordEvent(eventSource string, evt eventReport) {
-	select {
-	case <-er.ctx.Done():
-	case er.incomingReportChan <- report{eventSource, evt}:
-	}
-}
 
-func (er *EventRecorder) ingestReports() {
-	var queuedReports []report
-	var reportChan chan []report = nil
-	for {
-		select {
-		case <-er.ctx.Done():
-			return
-		// read incoming report
-		case report := <-er.incomingReportChan:
-			queuedReports = append(queuedReports, report)
-			if reportChan == nil {
-				// if queue had been idle, reactivate and set this report to publish next
-				reportChan = er.reportChan
+		switch ret := event.(type) {
+		case events.EventWithCandidates: // events.RetrievalEventCandidatesFound, events.RetrievalEventCandidatesFiltered:
+			protocols := ret.Protocols()
+			protocolStrings := make([]string, 0, len(protocols))
+			for _, protocol := range protocols {
+				protocolStrings = append(protocolStrings, protocol.String())
 			}
-		// send outgoing report when queue is active (when reportChan == nil, this path is never followed)
-		case reportChan <- queuedReports:
-			queuedReports = nil
-			reportChan = nil
-		}
-	}
-}
-
-// postReports picks reports from the reportChan and processes, them, an http
-// Client will be kept locally for each call to this function, one call per
-// desired parallelism can be made within goroutines, each with its own Client.
-func (er *EventRecorder) postReports() {
-	client := http.Client{Timeout: HttpTimeout}
-
-	for {
-		select {
-		case <-er.ctx.Done():
-			return
-		case reports := <-er.reportChan:
-			er.handleReports(client, reports)
```

Comment:
if you follow this to see the original `return`s were meaningless and didn't make the goroutine halt; maybe you should extract the http stuff back into functions again so it's neatly separated from the rest